[Refactor] Migration Module & 모듈 분리를 이용한 객체 의존성 분리

Marcorable opened the pull request:
## 📌 배경

close #110 

develop 빌드 가능 버전 필요

## 내용
1. #110 버전에서 빌드가능 버전을 위해 모호한 ViewModifier 및 View 주석처리
2. ShareExtension 의존성 주입 (Domain, UserInterface)
3. KakaoLogin 및 Update Article 동작 모듈 이동 (UserInterface -> Domain, API)
   > 이 부분에 관한 질문은 라이브 코딩으로라도 보완하겠습니다. 언제든 말씀 부탁드려요!
   > UserInterface (UI) 만 담당하는 Module에서 각각 역할을 담당하는 모듈로 이동했습니다.  
   > API 모듈 (네트워크 통신 기능)  /  Domain (데이터 관리)

혹시 빌드가 안되는 이슈나 에러가 발생 또는 유발하는 부분이 있으면 말씀 부탁드립니다.

file 변경 갯수가 너무 많아서 Commit 단위로 리뷰 부탁드립니다... ㅠㅠ

Marcorable committed:
chore: 빌드 에러 영역 주석처리

Marcorable committed:
feat: ShareExtension 의존성 주입

- UserInterface, Domain 추가

Marcorable committed:
refactor: LoginModel 생성 및 불필요한 파일 제거

- Domain
   - LoginModel 및 LoginModelProtocol 추가
- UserInterface
   - ModelContainer에 LoginModel 추가
- 불필요한 파일 제거
   - HomeDummyData

Marcorable committed:
refactor: Network 동작 의존성 분리 (KakaoLogin, ArticleModel)

- API 모듈: 중첩을 위한 파일제거
   - HTTPMethod.swift, NetworkRequest
- 의존성 분리를 위해 파일 이동
   - KakaoLoginRequest.swift, KakaoLoginResponse.swift
   - UploadArticleRequest.swift, UploadArticleResponse.swift
   - ErrorType.swift -> NetworkError.swift
- 의존성 분리를 위해 파일 제거
   - UpdateArticleAPI.swift
   - Router.swift
- LoginModel로 책임 변경
   - UserInfoService -> LoginModel

Marcorable committed:
refactor: API 모듈 Network 통신 기능 구현 및 Domain 모듈 요청 구현

- APIDetails: Network 통신 기능
- Domain/Models: 데이터 요청 (CRUD)
- ViewModel 내부 ModelContainer 추가

Marcorable committed:
refactor: Key 값 관리 및 초기화 시점 변경

- App 파일에서 Kakao 로그인 초기화

joseph704 commented on Targets/API/Sources/Request/KakaoLoginRequest.swift:
```diff
@@ -0,0 +1,38 @@
+
+import Foundation
+
+public struct KakaoLoginRequest: Encodable {
```
현재 통신 로직상 Encodable 프로토콜 제거해도 될 것 같은데, 혹시 추후에 필요하실 까요?

kipsong133 replied:
> 현재 통신 로직상 Encodable 프로토콜 제거해도 될 것 같은데, 혹시 추후에 필요하실 까요?

혹시, 통신로직상 필요 없다는 말씀이, 어떤 건지 여쭤봐도 될까요? HTTP Request Post Body 구성이 필요한데, 
Encodable 말고 다른 방법이 있는지 제가 모릅니다~!

Marcorable replied:
> > 현재 통신 로직상 Encodable 프로토콜 제거해도 될 것 같은데, 혹시 추후에 필요하실 까요?
> 
> 혹시, 통신로직상 필요 없다는 말씀이, 어떤 건지 여쭤봐도 될까요? HTTP Request Post Body 구성이 필요한데, Encodable 말고 다른 방법이 있는지 제가 모릅니다~!

저도 현재 작성된 코드에서는 

```swift
let json: [String: Any] = [
            "accessToken": "\(accessToken)",
            "vendor": "\(vendor)"
        ]
request.httpBody = try? JSONSerialization.data(withJSONObject: json, options: .prettyPrinted)
```

같은 형태로 Encoding 하고 있기 때문에 `Encodable`이 사용되지 않는다고 이해하고 있습니다.

저는 Codable로 JSON -> 객체, 객체 -> JSON 만 해보고 지금같은 URLRequest 구성은 처음인데
혹시 httpBody를 현재 구현형태 말고 다른 방식도 가능한가요?

joseph704 replied:
제가 말씀드린 부분이 용우님이 말씀주신 부분 맞습니다.
다시 생각해보니.. Encodable을 활용해 Request 객체를  httpBody에 넣는게 더 가시성이 좋을 것 같네요! 
Encodable제거가 아닌 Encodable을 사용을 위한 코드수정을 말씀드렸어야했는데 제가 이부분을 간과한 것 같습니다 ㅠ 

하기처럼 Encodable를 사용하는 방향으로 코드 제안 드려봅니다!
[UploadArticleRequest 코드 수정 제안](https://github.com/depromeet/ticlemoa-iOS/pull/111#discussion_r1061085835)
[KakaoLoginRequest 코드 수정 제안](https://github.com/depromeet/ticlemoa-iOS/pull/111#discussion_r1061086338)

joseph704 commented on Targets/API/Sources/Request/UploadArticleRequest.swift:
```diff
@@ -0,0 +1,60 @@
+
+import Foundation
+
+public struct UploadArticleRequest: Encodable {
```
Encodable 프로토콜 제거 제안

joseph704 commented on Targets/Domain/Sources/Model/ArticleModel.swift:
```diff
@@ -36,8 +36,25 @@ extension ArticleModel {
-        
+    public func update(_ item: Article) async {
+        do {
+            guard let data = try await api.uploadArticle(by: .init(
```
```suggestion
            guard let data = await api.uploadArticle(by: .init(
```
uploadArticle 메소드가 throws 하고 있지 않아 try 제거 해도 될 것 같습니다.

Marcorable replied:
오우 그러네요 놓쳤는데 감사합니다

joseph704 commented on Targets/DomainInterface/Sources/Entity/KakaoLoginResponse.swift:
```diff
@@ -10,4 +10,9 @@ import Foundation
 
 struct KakaoLoginResponse: Decodable {
```
접근제어를 public으로 해야할 것 같습니다! (Domain 모듈의 Model쪽에서 컴파일 에러 발생)

Marcorable replied:
아 수정하도록 하겠습니다..

joseph704 commented on Targets/DomainInterface/Sources/Entity/UploadArticleResponse.swift:
```diff
@@ -14,6 +14,8 @@
 // MARK: - UploadArticleResponse
+
+// TODO: Article 고민 필요
 struct UploadArticleResponse: Codable {
```
접근제어를 public으로 해야할 것 같습니다! (Domain 모듈의 Model쪽에서 컴파일 에러 발생)

joseph704 replied:
또한 Response에서는 encode할 일이 없기 떄문에, Codable -> Decodable로 바꾸면 좋을 것 같습니다!

Marcorable replied:
동의합니다 수정하겠습니다.

joseph704 commented on Targets/UserInterface/Sources/Login/LoginViewModel.swift:
```diff
@@ -7,14 +7,16 @@
+    
+    @EnvironmentObject var modelContainer: ModelContainer
     
     public func kakaoButtonDidTap() async throws -> Bool {
```
kakaoButtonDidTap 메소드 내부에서 예외처리가 Bool값으로 잘 반환하셨기 때문에, 
throws를 제거해도 될 것 같은데 어떠신가요?

kipsong133 replied:
> kakaoButtonDidTap 메소드 내부에서 예외처리가 Bool값으로 잘 반환하셨기 때문에, throws를 제거해도 될 것 같은데 어떠신가요?

네 throws 를 제거해도 되지만, async await 가 있는 스텍 중에서(호출된 로직) 만약 throw 가 있다면,
throw 가 잇어야 컴파일 에러가 없을 겁니다.(이부분은 제가 작성한지 조금 되서 저도 직접 컴파일 해보겠습니다. 확인해주셔서 감사합니다!)

사실, 말씀해주신 것처럼, throws 쓰는 기준이나 bool 쓰는 기준을 제가 공유드렸어야했는데, 
후딱한다고 일단 작성만 했습니다. 
개인적으로 throws 는 성공과실패를 떠나서 호출자체를 실패하는 경우를 나타내고 싶었고,
bool 의 경우, 호출은 되었으나 불충분 조건으로 호출된 경우를 했었는데, 

아이디어 공유해주시면 함께 맞추면 좋을 듯 합니다.
일차적으로는 말씀해주신 throws 제거하는것으로 해볼게요!

Marcorable replied:
우성님 말씀처럼 
통신 시도에 대한 처리를 throws 로 Error 처리하는게 좋은 방향 같습니다.

제가 수정하다가 생긴 부분인데 UI에서는 async Bool로 처리하는게 좋을것 같아서 변경하는 과정에서 생겼습니다.

통신 부분에서 throws를 해야하기때문에 위 로직에는 필요하긴 한데..
이 부분이 UserInterface 모듈에 있는게 이 로직의 성격이 좀 다르다고 생각합니다..
KakaoSDK 자체가 모두 LoginModel로 이동하면 어떨까요?

sustainable-git replied:
저도 LoginModel로 이동하는게 좋다고 생각해요!
ViewModel은 Task{} 함수로 비동기를 실행하고,
그 결과를 ViewModel의 @Published property를 변경해서서 View에 알려주는게 맞지 않을까요?

joseph704 replied:
우성님 말씀 감사합니다!
저도 KaKaoSDK 로직은 LoginModel에서 처리하고, ViewModel에서 결과값을 처리하는 것이 나을 것 같습니다.

keeplo replied:
@sustainable-git 

`@Published` 프로퍼티를 처음 생각했었는데.. 
Model 간에 전달도 어렵고, UserInteface 부터 API까지 전달해서 가는것도 어색할거 같더라구요
`UserDefault` 로 저장하면 어떨까요? 
좀 더 접근이 용이하고 한번 로그인 완료한 계정을 나타내기에도 좋지 않을까 생각됩니다!

Marcorable replied:
@sustainable-git

저도 @Published 프로퍼티를 처음 생각했었는데..
Model 간에 전달도 어렵고, UserInteface 부터 API까지 전달해서 가는것도 어색할거 같더라구요
UserDefault 로 저장하면 어떨까요?
좀 더 접근이 용이하고 한번 로그인 완료한 계정을 나타내기에도 좋지 않을까 생각됩니다!

joseph704 reviewed (changes requested):
고생하셨습니다! 👍👍

kipsong133 reviewed (approved):
확인했습니다.
늦은밤까지 고생 많으셨습니다. 감사합니다!

Marcorable committed:
refactor: Response 모듈 위치 변경 및 Domain Entity 변경

- Response 개체 모듈 이동 DomainInterface -> API
- ArticleData - UploadArticleRequest 호환 되도록 수정
- 마이그레이션 (Router -> TiclemoaAPI) 마친 개체 삭제

Marcorable committed:
refactor: DomainInterface 수정 누락내용 반영

- protocol Article 수정

sustainable-git commented on Targets/API/Sources/APIDetails.swift:
```diff
@@ -25,6 +34,48 @@ public struct TiclemoaAPI {
+    public func kakaoLogin(by request: KakaoLoginRequest) async -> Data? {
+        let urlRequest = request.makeURLRequest(by: baseURL)
+        
+        do {
+            let (data, response) = try await URLSession.shared.data(for: urlRequest)
+            let httpResponse = response as? HTTPURLResponse
+            print("DEBUG: ", httpResponse!)
+            
+            guard (response as? HTTPURLResponse)?.statusCode == 200 || (response as? HTTPURLResponse)?.statusCode == 201 else {
+                return nil
+            }
+            print("DEUBG: \(response.description)")
+            return data
+        } catch {
+            return nil
+        }
```
Error 처리를 여기서 해주면 API를 사용하는 단에서는 어떤 에러가 나는지 확인이 안됩니다.
따라서 유저에게 404인지 503인지 알려줄 수가 없습니다.
throws function으로 만드는게 어떨까요?
최소한 Logger를 사용해서 어떤 log가 발생했는지 기록이라도 해야할 것 같아요.

Marcorable replied:
말씀하신 방향이 좋을것 같습니다.
UserInterface 까지 문자열로 라도 전달되도록 구현을 바꿔봐야겠네요!
PR 체이닝으로 작업중인 부분있는데 해당 부분 고려해서 수정해보겠습니다

joseph704 commented on Targets/API/Sources/Request/UploadArticleRequest.swift:
```diff
@@ -0,0 +1,60 @@
+        request.httpMethod = HTTPMethod.post.rawValue
+        request.addValue("application/json", forHTTPHeaderField: "Content-Type")
+        
+        let json: [String: Any] = [
+            "content"   : content,
+            "userId"    : userId,
+            "title"     : title,
+            "url"       : url,
+            "isPublic"  : isPublic,
+            "tagIds"    : tagIds,
+        ]
+        request.addValue("Bearer \(accessToken)", forHTTPHeaderField: "Authorization")
+        request.httpBody = try? JSONSerialization.data(withJSONObject: json, options: .prettyPrinted)
```
```suggestion
       request.addValue("Bearer \(accessToken)", forHTTPHeaderField: "Authorization")
        let jsonData = try? JSONEncoder().encode(self)
        request.httpBody = jsonData
```

joseph704 replied:
[참고](https://github.com/depromeet/ticlemoa-iOS/pull/111#discussion_r1061087197)

Marcorable replied:
혹시 `.prettyPrinted` 옵션을 제가 몰라서 그런데
그 부분은 상관없이 기대대로 동작되나요?

Marcorable replied:
자답입니다 [사용 예시](https://zrr.kr/q70C) 

다른 용도의 처리는 없어서 적용하겠습니다!
제가 체이닝으로 작업중인 PR 있었는데 거기에 반영하겠습니다!

joseph704 commented on Targets/API/Sources/Request/KakaoLoginRequest.swift:
```diff
@@ -0,0 +1,38 @@
+        request.httpMethod = HTTPMethod.post.rawValue
+        request.addValue("application/json", forHTTPHeaderField: "Content-Type")
+        
+        let json: [String: Any] = [
+            "accessToken": "\(accessToken)",
+            "vendor": "\(vendor)"
+        ]
+        request.httpBody = try? JSONSerialization.data(withJSONObject: json, options: .prettyPrinted)
```
```suggestion
        let jsonData = try? JSONEncoder().encode(self)
        request.httpBody = jsonData
```

joseph704 replied:
[참고](https://github.com/depromeet/ticlemoa-iOS/pull/111#discussion_r1061087197)

Marcorable committed:
refactor: Build를 위해서 수정

- Home/Network/GetArticle/ResponseArticle.swift
- public 추가
- bundleId 모듈 이름 추가
